---
layout: default
title: How to Cite
nav_order: 4
permalink: /how-to-cite
---

# How to cite
Tice AK*, Žihala D*, Pánek T*, Jones RE", Salomaki E", Nenarokov S, Burki F, Eliáš M, Eme L, Roger AJ, Rokas A, Shen X, Strassert JFH, Kolísko M, Brown MW. 2021. PhyloFisher: A phylogenomic package for resolving eukaryotic relationships. [PLoS Biology. 19(8): e3001365](https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.3001365) * & " =Contributed equally

# Third Party Software/Data Included
*If you use PhyloFisher, please also cite these invaluable resources.*

ASTRAL-III (Zhang et al., 2018) - Zhang C, Rabiee M, Sayyari E, Mirarab S. ASTRAL-III: polynomial time species tree reconstruction from partially resolved gene trees. BMC Bioinformatics. 2018;19: 153. doi:10.1186/s12859-018-2129-y

BLAST v. 2.9.0 (Camacho et al., 2009) - Camacho, C., Coulouris, G., Avagyan, V., Ma, N., Papadopoulos, J., Bealer, K., & Madden, T. L. (2009). BLAST+: Architecture and applications. BMC Bioinformatics, 10(1), 421. https://doi.org/10.1186/1471-2105-10-421

BMGE v. 1.1.2 (Criscuolo & Gribaldo, 2010) - Criscuolo, A., & Gribaldo, S. (2010). BMGE (Block Mapping and Gathering with Entropy): A new software for selection of phylogenetic informative regions from multiple sequence alignments. BMC Evolutionary Biology, 10(1), 210. https://doi.org/10.1186/1471-2148-10-210

CD-HIT v. 4.8.1 (Fu et al., 2012) - Fu, L., Niu, B., Zhu, Z., Wu, S., & Li, W. (2012). CD-HIT: accelerated for clustering the next-generation sequencing data. Bioinformatics, 28(23), 3150–3152. https://doi.org/10.1093/bioinformatics/bts565

DIAMOND v. 09.24 (Buchfink et al., 2015) - Buchfink, B., Xie, C., & Huson, D. H. (2015). Fast and sensitive protein alignment using DIAMOND. Nature Methods, 12(1), 59–60. https://doi.org/10.1038/nmeth.3176

DIST_EST v.1.0 (Susko et al., 2003) - Susko, E., & Roger, A. J. (2007). On Reduced Amino Acid Alphabets for Phylogenetic Inference. Molecular Biology and Evolution, 24(9), 2139–2150. https://doi.org/10.1093/molbev/msm144

DIVVIER v. 1.01 (Ali et al., 2019) - Ali, R. H., Bogusz, M., & Whelan, S. (2019). Identifying Clusters of High Confidence Homologies in Multiple Sequence Alignments. Molecular Biology and Evolution, 36(10), 2340–2351. https://doi.org/10.1093/molbev/msz142

ETE3 3.1.1 (Huerta-Cepas et al., 2016) - Huerta-Cepas, J., Serra, F., & Bork, P. (2016). ETE 3: Reconstruction, Analysis, and Visualization of Phylogenomic Data. Molecular Biology and Evolution, 33(6), 1635–1638. https://doi.org/10.1093/molbev/msw046

FastTree v. 2.1.11 (Price et al., 2010) - Price, M. N., Dehal, P. S., & Arkin, A. P. (2010). FastTree 2—Approximately Maximum-Likelihood Trees for Large Alignments. PLoS ONE, 5(3), 1–10. a9h.

HMMER v. 3.2.1 (Mistry et al., 2013) - Mistry, J., Finn, R. D., Eddy, S. R., Bateman, A., & Punta, M. (2013). Challenges in homology search: HMMER3 and convergent evolution of coiled-coil regions. Nucleic Acids Research, 41(12), e121–e121. https://doi.org/10.1093/nar/gkt263

MAFFT v.7.455 (Katoh & Standley, 2013) - Katoh, K., & Standley, D. M. (2013). MAFFT Multiple Sequence Alignment Software Version 7: Improvements in Performance and Usability. Molecular Biology and Evolution, 30(4), 772–780. https://doi.org/10.1093/molbev/mst010

MAMMaL v1.1.1 (Susko et al., 2018) - Susko, E., Lincker, L. and Roger, A.J. (2018). Accelerated Estimation of Frequency Classes in Site-heterogeneous Profile Mixture Models. Molecular Biology and Evolution. 9:1266-1283.

OrthoMCL v 5.0 (Chen et al., 2006) - Chen, F., Mackey, A. J., Stoeckert, C. J., Jr, & Roos, D. S. (2006). OrthoMCL-DB: querying a comprehensive multi-species collection of ortholog groups. Nucleic Acids Research, 34(suppl_1), D363–D368. https://doi.org/10.1093/nar/gkj123

PREQUAL v. 1.02 (Whelan et al., 2018) - Whelan, S., Irisarri, I., & Burki, F. (2018). PREQUAL: detecting non-homologous characters in sets of unaligned homologous sequences. Bioinformatics, 34(22), 3929–3930. https://doi.org/10.1093/bioinformatics/bty448

RAxML v. 8.2.12 (Stamatakis, 2014) - Stamatakis, A. (2014). RAxML version 8: A tool for phylogenetic analysis and post-analysis of large phylogenies. Bioinformatics, 30(9), 1312–1313. https://doi.org/10.1093/bioinformatics/btu033

trimAl v.1.4.rev15 (Capella-Gutiérrez et al., 2009) - Capella-Gutiérrez, S., Silla-Martínez, J. M., & Gabaldón, T. (2009). trimAl: A tool for automated alignment trimming in large-scale phylogenetic analyses. Bioinformatics, 25(15), 1972–1973. https://doi.org/10.1093/bioinformatics/btp348